dhall/src/semantics/nze/nir.rs

```diff
@@ -138,6 +138,24 @@ impl Nir {
         self.0.kind()
     }
 
+    /// The contents of a `Nir` are immutable and shared. If however we happen to be the sole
+    /// owners, we can mutate it directly. Otherwise, this clones the internal value first.
+    pub fn kind_mut(&mut self) -> &mut NirKind {
+        if Rc::get_mut(&mut self.0).is_none() {
+            // Clone self
+            let kind = self.kind().clone();
+            *self = Nir::from_kind(kind);
+        }
+        Rc::get_mut(&mut self.0).unwrap().kind_mut()
```

Comment:
Would [`Rc::make_mut`](https://doc.rust-lang.org/std/rc/struct.Rc.html#method.make_mut) work here?

Reply:
I considered it, but I would have had to make `NirInternal` `Clone`, which I prefer to avoid in order to avoid spurious clones. I checked the impls and it doesn't look much less efficient like this

Reply:
Granted, I'm not sure it makes much difference to make NirInternal `Clone`. I made that choice a while ago when that bit of the code was way more conplicated

Reply:
Actually you made me realize I could easily get rid of `NirInternal`. That's easier and I can use `make_mut`